### PR TITLE
[SPARK-23827] [SS] StreamingJoinExec should ensure that input data is partitioned into specific number of partitions

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

Currently, the requiredChildDistribution does not specify the partitions. This can cause the weird corner cases where the child's distribution is `SinglePartition` which satisfies the required distribution of `ClusterDistribution(no-num-partition-requirement)`, thus eliminating the shuffle needed to repartition input data into the required number of partitions (i.e. same as state stores). That can lead to "file not found" errors on the state store delta files as the micro-batch-with-no-shuffle will not run certain tasks and therefore not generate the expected state store delta files.

This PR adds the required constraint on the number of partitions.

## How was this patch tested?
Modified test harness to always check that ANY stateful operator should have a constraint on the number of partitions. As part of that, the existing opt-in checks on child output partitioning were removed, as they are redundant.

### CODE REVIEW

**Comment by tdas** [comment]
@brkyvz @zsxwing can one of you take a look?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala [hunk 1]
```diff
@@ -444,6 +445,26 @@ trait StreamTest extends QueryTest with SharedSQLContext with TimeLimits with Be
...
+        executedPlan.collect { case s: StatefulOperator => s }.foreach { s =>
+          assert(s.stateInfo.isDefined)
+          assert(s.stateInfo.get.numPartitions >= 1)
+
+          s.requiredChildDistribution.foreach { d =>
+            withClue(s"$s specifies incorrect # partitions in requiredChildDistribution $d") {
+              assert(d.requiredNumPartitions.isDefined)
+              assert(d.requiredNumPartitions.get >= 1)
+              if (d != AllTuples) {
+                assert(d.requiredNumPartitions.get == s.stateInfo.get.numPartitions)
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala]: can you also verify that this is equal to the number of partitions in the metadata?

**Review by brkyvz — APPROVED** [review]
LGTM. This is great. Left one comment

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala [hunk 2]
```diff
@@ -450,8 +450,8 @@ trait StreamTest extends QueryTest with SharedSQLContext with TimeLimits with Be
         // This can often catch hard to debug errors when developing stateful operators
         val executedPlan = currentStream.lastExecution.executedPlan
         executedPlan.collect { case s: StatefulOperator => s }.foreach { s =>
-          assert(s.stateInfo.isDefined)
-          assert(s.stateInfo.get.numPartitions >= 1)
+          assert(
+            s.stateInfo.map(_.numPartitions).contains(currentStream.lastExecution.numStateStores))
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala]: why this change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala [hunk 3]
```diff
@@ -167,7 +167,8 @@ case class StreamingSymmetricHashJoinExec(
   val nullRight = new GenericInternalRow(right.output.map(_.withNullability(true)).length)
 
   override def requiredChildDistribution: Seq[Distribution] =
-    ClusteredDistribution(leftKeys) :: ClusteredDistribution(rightKeys) :: Nil
+    ClusteredDistribution(leftKeys, stateInfo.map(_.numPartitions)) ::
+      ClusteredDistribution(rightKeys, stateInfo.map(_.numPartitions)) :: Nil
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala]: +1

**Comment by tdas** [comment]
Started more tests to test for flakiness.

**Review by tedyu — APPROVED** [review]
